### PR TITLE
feat: Add .dryrun Command Support 

### LINKED ISSUE #156: feat: dryRun support in aos
The ability to use dryRuns to query Processes without having to send a message.

```
.dryrun Balances
```

Basically, command then expression

```
.dryrun {Lua Expression}
```

### PR DESCRIPTION
**Pull request by mayurmarvel**
### Description
This pull request addresses issue #156 by introducing support for the `.dryrun` command in the `aos` project. Below is a summary of the changes made:

#### Files Modified / Created:
- `/src/index.js`: Added support for `.dryrun` command.
- `/src/commands/dryrun.js`: Implemented the functionality for the `.dryrun` command.
- `/src/commands/dryrun.md`: Created a usage guide for the `.dryrun` command.
- `README.md`: Updated commands list to include `.dryrun`.
- `/src/services/help.js`: Added `.dryrun` to the help list.

**For the full usage guide and detailed changes, please refer to** `/src/commands/dryrun.md`.

### Command Example
      .dryrun ({ Target = 'FgU-RiEaLuC__SHZnI9pSIa_ZI8o-8hUVG9nPJvs92k', Action = 'Balance' })

      .dryrun {Balance}

      .dryrun -r.Messages[0] {Balance}

### Flags List:
- `-p`: Specifies Process ID.
- `-d`: Specifies data.
- `-r`: Accesses the Result Object.
- `-v`: Views the DryRun Parameters.

### Functions:
- `dryrun`: Executes a dry run to query Processes without sending a message.

### Testing
Performed manual testing to ensure the functionality of the `.dryrun` command in various scenarios.

### Features That Can Be Added
- A loading spinner can be added if needed.

### Related Issue
Closes #156

### CODE REVIEW

**Review by twilson63 — APPROVED** [review]

File: src/commands/dryrun.js [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+function checkForVerboseFlag(input, dryrunParams) {
+    const regex = /(^|\s)-v($|\s)/; // Regular expression to match -v preceded and followed by whitespace or string boundary
+
+    if (regex.test(input)) {
+        console.log("dryrunParams:", dryrunParams);
+    }
+}
+
+
+const getCustomProcessId = (input, pid) => {
```
Comment (allquantor) [inline src/commands/dryrun.js:108]: Slightly cleaner. 
```
const getCustomProcessId = (input, pid) => {
  const match = input.match(/-p=(["'])(.*?)\1/);
  if (!match) return null;
  const customProcessId = match[2] === "ao.id" ? pid : match[2];
  if (customProcessId.length !== 43) throw new Error("Invalid Process Id");
  return customProcessId;
};
```

File: src/commands/dryrun.js [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+      ...(id && { Id: id }),
+      ...(anchor && { anchor: anchor }),
+    };
+
+    checkForVerboseFlag(input, dryrunParams);
+
+    const result = await dryrun(dryrunParams);
+    return result;
+  } catch (error) {
+    console.error(chalk.red("Error Performing DryRun:", error.message));
```
Comment (allquantor) [inline src/commands/dryrun.js:95]: Would it not be right to propagate the error up in context? Why swallow it here.

File: src/commands/dryrun.js [hunk 3]
```diff
@@ -0,0 +1,176 @@
+import { dryrun } from "@permaweb/aoconnect";
+import chalk from "chalk";
+
+export async function performDryRun(input, pid, data, owner, id, anchor) {
+  await processInput(input, pid, data, owner, id, anchor);
+}
+
+async function extractTags(input, pid) {
```
[the remark this answers is not in the document]
Comment (allquantor) [reply src/commands/dryrun.js:8]: P.S if target can be appended you can get rid of else al in all 

```
if (!input.includes("{")) throw new Error("Tags not specified");

  const content = input.match(/\{(.*?)\}/)?.[1];
  if (!content) throw new Error("Invalid Syntax Usage");

  return content.split(",").map(pair => {
    const [key, rawValue] = pair.includes("=") ? pair.split("=") : ["Action", pair];
    const value = rawValue.trim().replace(/^['"](.*)['"]$/, "$1").replace(/^ao\.id$/, pid);
    return { name: key.trim(), value };
  }).concat({ name: "Target", value: pid });
  ```

**Review by allquantor — CHANGES_REQUESTED** [review]
Added a couple of comments. 
I would re-iterate to reduce overall complexity. A lot of stuff to simplify. 
Good job

File: README.md [hunk 4]
```diff
@@ -70,12 +70,13 @@ When running the console, you can type `dot` commands to instruct the console to
...
+| `.dryrun {tags}`         | This command will allow you to perform dryrun.                                                                                               |
 | `.exit`                  | This command will exit you console, but you can also do `Ctrl-C` or `Ctrl-D`                                                                 |
 
 ## License
 
 The ao and aos codebases are offered under the BSL 1.1 license for the duration
 of the testnet period. After the testnet phase is over, the code will be made
 available under either a new
-[evolutionary forking](https://arweave.medium.com/arweave-is-an-evolutionary-protocol-e072f5e69eaa)
+[evolutionary forking](https://arweave.medium.com/arweave-is-an-evolutionary-protocol-e072f5e69eaa)s
```
[the remark this answers is not in the document]
Comment (mayurmarvel) [reply README.md:81]: Sorry, that's a typo.

**Comment by twilson63** [comment]
🤔 I think this can be a simpler fix, if you spawn the process and you are not the owner, you can open the process in (Read-Only) mode by default, which means, that instead of sending messages, you send `dryrun`s - so you will not be able to perform any write capable functions, because you are not the Owner, but you will be able to inspect the process and visually view the state?

@allquantor @mayurmarvel - What do you think?

For example, if I call `aos {pid}` and I don't have `Eval` privileges, then aos does `dryrun` be default?

**Pull request closed without merging**